[Hot/Cold Fix] Fix the implementation of the "minecraft:register" and "minecraft:unregister" payloads

marchermans opened the pull request:
### Update on Network Protocol
We are still in negotiations on the new cross modloader networking protocol, however as of not this has not born fruit. The overlap in the discussions we are currently holding, can be covered by the legacy Dinnerbone protocol, for which this PR restores support.

### Fixes:
- Implement the Dinnerbone protocol. 

### Tested:
- Vanilla Client -> Neo Server
- Neo Client -> Vanilla Server
- Neo Client -> Neo Server
- Neo Client -> Fabric Server
- Fabric Client -> Neo Server
- Neo Client -> Paper Server
- Neo Client -> BungeeCord Server

### Test mods:
https://gist.github.com/marchermans/310c60d472ca4a82063439ae7d007f3f
The code in the gist registers a delayed listener on Fabric.
If it works you should see a reply from the server every second.

marchermans committed:
Initial hotfix

marchermans committed:
Implemented core connection logic with connection types.

marchermans committed:
Implement best practices for connection detection: use isConnection(ResLoc)

XFactHD commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -859,6 +934,163 @@ public void configureMockConnection(final Connection connection) {
+    private void onMinecraftRegister(Set<ResourceLocation> resourceLocations, Connection connection) {
+        final Set<ResourceLocation> payloadSetup = connection.channel().attr(ATTRIBUTE_ADHOC_CHANNELS).get();
+        if (payloadSetup == null) {
+            return;
```
When the ad-hoc channel attribute is not set yet, the attribute should be initialized with the incoming set of new channels instead of the set being discarded

XFactHD commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -859,6 +934,163 @@ public void configureMockConnection(final Connection connection) {
+    private void onMinecraftUnregister(Set<ResourceLocation> resourceLocations, Connection connection) {
+        final Set<ResourceLocation> payloadSetup = connection.channel().attr(ATTRIBUTE_ADHOC_CHANNELS).get();
+        if (payloadSetup == null) {
+            connection.channel().attr(ATTRIBUTE_ADHOC_CHANNELS).set(resourceLocations);
```
When the ad-hoc channel attribute is not set yet, the incoming set of channels-to-be-forgotten should be discarded instead of the attribute being initialized with that set

XFactHD commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -859,6 +934,163 @@ public void configureMockConnection(final Connection connection) {
+    }
+
+    public Set<ResourceLocation> getInitialServerUnregisterChannels() {
+        final ImmutableSet.Builder<ResourceLocation> nowListeningOn = ImmutableSet.builder();
```
This variable name is "inverted" for a set of channels which won't be listened for anymore

XFactHD commented on patches/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java.patch:
```diff
@@ -1,23 +1,40 @@
+ import net.minecraft.world.flag.FeatureFlags;
+ import net.neoforged.api.distmarker.Dist;
+ import net.neoforged.api.distmarker.OnlyIn;
++import net.neoforged.neoforge.network.connection.ConnectionType;
```
Undesired import

XFactHD commented on patches/net/minecraft/client/multiplayer/ClientPacketListener.java.patch:
```diff
@@ -1,18 +1,50 @@
 --- a/net/minecraft/client/multiplayer/ClientPacketListener.java
 +++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
+@@ -6,7 +_,6 @@
```
Undesired import changes

XFactHD commented on patches/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch:
```diff
@@ -1,18 +1,29 @@
+ import net.minecraft.server.players.PlayerList;
+ import net.minecraft.tags.TagNetworkSerialization;
+ import net.minecraft.world.flag.FeatureFlags;
++import net.neoforged.neoforge.network.connection.ConnectionType;
```
Undesired import

XFactHD commented on patches/net/minecraft/server/network/ServerGamePacketListenerImpl.java.patch:
```diff
@@ -1,5 +1,21 @@
 --- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
 +++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+@@ -6,7 +_,6 @@
```
Undesired import changes

XFactHD commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -6,20 +6,18 @@
-import java.util.Optional;
-import java.util.Set;
+
+import java.util.*;
```
No wildcard imports and the import order doesn't match the rules. Needs a spotless run

XFactHD reviewed (changes requested):
Would potentially be interesting to check a client with these changes against weirder setups like ViaVersion (does Hypixel use that or something else?)

XFactHD commented:
While we're at it: we should fix the error messages for connection failures.

- [This](https://github.com/neoforged/NeoForge/blob/ddaf54db080088636da3efa8d0f7ae752107f381/src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java#L529) and [this](https://github.com/neoforged/NeoForge/blob/ddaf54db080088636da3efa8d0f7ae752107f381/src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java#L541) need to be a literals since the vanilla client doesn't have our translation strings
- [This](https://github.com/neoforged/NeoForge/blob/ddaf54db080088636da3efa8d0f7ae752107f381/src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java#L718) and [this](https://github.com/neoforged/NeoForge/blob/ddaf54db080088636da3efa8d0f7ae752107f381/src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java#L730) talks about a vanilla client attempting to connect to a NF server. This code path is involved with the other direction though (NF client with non-optional channels connecting to a vanilla server)
- The disconnection message always suggests installing a certain NF version, even though that is neither required in order to satisfy NF's own payloads (NF's own payloads are versioned by the MC version part of the NF version) nor is it correct in most cases (i.e. missing mod or mismatched mod versions)

pupnewfster commented:
Can those that are can show for vanilla use `Component#translatableWithFallback`? Or are they only ever shown to vanilla clients and never potentially neo ones who may have the translation string?

XFactHD commented:
They are only ever shown to vanilla clients

Technici4n committed:
Remove patch imports and spotless

Technici4n committed:
Simplify ad-hoc channel handling and move logging out of modder-callable isConnected

marchermans committed:
Fix the configuration not loading when running against a client that does not send the brandpayload.

Tested connections:
Client -> Server
Neo -> Neo
Vanilla -> Neo
Neo -> Vanilla
Neo -> Paper
Neo -> Bungecord (HyPixel)
Fabric -> Neo (Failed, Fabrics protocol does not work)
Neo -> Fabric

marchermans committed:
Patch fabric adhoc logic.

marchermans committed:
Gen patches and spotless

marchermans committed:
Merge branch '1.20.x' into hotfix/1.20.x/networking

marchermans committed:
Fix licenses

marchermans committed:
Merge remote-tracking branch 'origin/hotfix/1.20.x/networking' into hotfix/1.20.x/networking

XFactHD commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -686,7 +804,7 @@ public boolean onVanillaNetworkConnectionEstablished(ClientConfigurationPacketLi
         //Negotiation failed. Disconnect the client.
         if (!configurationNegotiationResult.success()) {
-            sender.getConnection().disconnect(Component.translatable("neoforge.network.negotiation.failure.vanilla.client.not_supported", NeoForgeVersion.getVersion()));
+            sender.getConnection().disconnect(Component.translatableWithFallback("neoforge.network.negotiation.failure.vanilla.client.not_supported", "You are trying to connect to a server that is running NeoForge, but you are not. Please install NeoForge Version: %s to connect to this server.", NeoForgeVersion.getVersion()));
```
Both the translation key and the fallback message are backwards. This codepath is invoked when a NF client connects to a vanilla server and fails

XFactHD commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -698,14 +816,22 @@ public boolean onVanillaNetworkConnectionEstablished(ClientConfigurationPacketLi
         //Negotiation failed. Disconnect the client.
         if (!playNegotiationResult.success()) {
-            sender.getConnection().disconnect(Component.translatable("neoforge.network.negotiation.failure.vanilla.client.not_supported", NeoForgeVersion.getVersion()));
+            sender.getConnection().disconnect(Component.translatableWithFallback("neoforge.network.negotiation.failure.vanilla.client.not_supported", "You are trying to connect to a server that is running NeoForge, but you are not. Please install NeoForge Version: %s to connect to this server.", NeoForgeVersion.getVersion()));
```
Both the translation key and the fallback message are backwards. This codepath is invoked when a NF client connects to a vanilla server and fails

XFactHD commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -607,12 +653,79 @@ public boolean canSendPacket(Packet<?> packet, ClientCommonPacketListener listen
+     * @param flow The flow of the packet.
+     * @return True if the packet is ad hoc readable, false otherwise.
+     */
+    private boolean isAdhocPlayReadable(ResourceLocation id, PacketFlow flow) {
```
Should be named `isAdhocPlayChannelReadable()` to match the other method

Technici4n commented on src/main/java/net/neoforged/neoforge/common/extensions/IFriendlyByteBufExtension.java:
```diff
@@ -175,4 +177,19 @@ default <K, V> void writeMap(Map<K, V> map, FriendlyByteBuf.Writer<K> keyWriter,
         });
     }
+
+    default <T> void writeSplit(Set<T> newChannels, Function<T, String> writer, String separator) {
```
This should be removed.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/extensions/IServerCommonPacketListenerExtension.java:
```diff
@@ -106,4 +107,9 @@ default boolean isConnected(final ResourceLocation payloadId) {
+    /**
+     * {@return the connection type of the connection}
+     */
+    ConnectionType getConnectionType();
```
Do we want to deprecate isVanillaConnection in favor of this method?

Technici4n replied:
I also don't like that `isVanillaConnection` gets its information from a different source than `getConnectionType`. We should store the connection type in the channel once we know it.

Technici4n commented on src/main/java/net/neoforged/neoforge/network/connection/ConnectionType.java:
```diff
@@ -0,0 +1,36 @@
+
+public enum ConnectionType {
+    VANILLA,
+    MODDED,
```
Should be renamed to OTHER, ideally.

Technici4n commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -79,6 +86,7 @@ public class NetworkRegistry {
 
     private static final AttributeKey<NetworkPayloadSetup> ATTRIBUTE_PAYLOAD_SETUP = AttributeKey.valueOf("neoforge:payload_setup");
+    private static final AttributeKey<Set<ResourceLocation>> ATTRIBUTE_ADHOC_CHANNELS = AttributeKey.valueOf("neoforge:adhoc_channels");
     private static final AttributeKey<Boolean> ATTRIBUTE_IS_MODDED_CONNECTION = AttributeKey.valueOf("neoforge:is_modded_connection");
```
This should be replaced by a ConnectionType. Then we could also directly get the ConnectionType from the channel, and not need a second parameter for the packet filters.

Technici4n commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -204,9 +221,9 @@ public FriendlyByteBuf.Reader<? extends CustomPacketPayload> getReader(ResourceL
                     final PacketListener listener = connection.getPacketListener();
                     if (listener instanceof ServerGamePacketListener serverListener) {
-                        serverListener.disconnect(Component.translatable("neoforge.network.invalid_flow", flow));
+                        serverListener.disconnect(Component.translatableWithFallback("neoforge.network.invalid_flow", "Failed to process a payload that was send with an invalid flow: %s", flow));
```
Not a fan of the translation duplication. The server should have translations loaded, so we can query the lang on the server instead of hardcoding its value.

marchermans replied:
Actually it does not.
We do not load the translations anymore on the server side, I also thought we did, but that is not the case anymore so this is currently the best we can do.

We already discussed internally that we should indeed refactor this and probably generate our en_us file using a data generator, so that this gets cleaned.

Technici4n commented on patches/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java.patch:
```diff
@@ -4,67 +4,95 @@
 +        }
-+        if (!this.isModdedConnection && p_295411_ instanceof net.minecraft.network.protocol.common.custom.BrandPayload) {
-+            this.isVanillaConnection = true;
++        if (!this.connectionType.isNeoForge() && p_295411_ instanceof net.minecraft.network.protocol.common.custom.BrandPayload) {
```
I don't like relying on Brand. We should just check the connection type when we receive Ping.

marchermans replied:
Sadly it is what is.
We discussed this very early on even before the original port, but during testing and processing this was the best we could come up with, we tried the approach on just relying on the ping packet on the client side, but it simply was not stable enough.

Technici4n replied:
Fair enough

Technici4n reviewed (changes requested):
Should work, but I think the code can be improved a little bit.

Technici4n commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -856,7 +958,149 @@ public void configureMockConnection(final Connection connection) {
+        final ImmutableSet.Builder<ResourceLocation> nowListeningOn = ImmutableSet.builder();
+        nowListeningOn.add(MinecraftRegisterPayload.ID);
+        nowListeningOn.add(MinecraftUnregisterPayload.ID);
+        if (!setup.vanilla()) {
```
The `vanilla` boolean should be removed from the `NetworkPayloadSetup`. This information is already accessible via the connection and/or the listener.

Technici4n commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -518,7 +543,7 @@ public boolean onVanillaConnectionDetectedAtServer(ServerConfigurationPacketList
 
```
We probably want to rename `onVanillaConnectionDetectedAtServer` to indicate that it also works for non-neoforge connection type. For example `onVanillaOrOtherConnectionDetectedAtServer` (assuming `ConnectionType.MODDED` is renamed too).

marchermans committed:
Address required fixes.

marchermans committed:
Remove the vanilla property from the channel setup.

Technici4n commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -799,12 +800,12 @@ public boolean onVanillaNetworkConnectionEstablished(ClientConfigurationPacketLi
         //Negotiation failed. Disconnect the client.
         if (!configurationNegotiationResult.success()) {
-            sender.getConnection().disconnect(Component.translatableWithFallback("neoforge.network.negotiation.failure.vanilla.client.not_supported", "You are trying to connect to a server that is running NeoForge, but you are not. Please install NeoForge Version: %s to connect to this server.", NeoForgeVersion.getVersion()));
+            sender.getConnection().disconnect(Component.translatableWithFallback("neoforge.network.negotiation.failure.vanilla.server.not_supported", "You are trying to connect to a server that is not running NeoForge, but you are not. A connection could not be established.", NeoForgeVersion.getVersion()));
```
", but you are not." is incorrect I think.

Maybe something like ", but you have mods that require it."

XFactHD replied:
The latter makes a lot more sense. This of course also applies to the line in the lang file

marchermans committed:
Remove the vanilla property from the channel setup.

Technici4n commented on src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java:
```diff
@@ -816,7 +817,7 @@ public boolean onVanillaNetworkConnectionEstablished(ClientConfigurationPacketLi
         //Negotiation failed. Disconnect the client.
         if (!playNegotiationResult.success()) {
-            sender.getConnection().disconnect(Component.translatableWithFallback("neoforge.network.negotiation.failure.vanilla.client.not_supported", "You are trying to connect to a server that is running NeoForge, but you are not. Please install NeoForge Version: %s to connect to this server.", NeoForgeVersion.getVersion()));
+            sender.getConnection().disconnect(Component.translatableWithFallback("neoforge.network.negotiation.failure.vanilla.server.not_supported", "You are trying to connect to a server that is not running NeoForge, but you are not. A connection could not be established.", NeoForgeVersion.getVersion()));
```
Same lang comment here.

Potentially we could reduce the duplication by first negotiating configuration and play, and then check `if (!configurationNegotiationResult.success() || !playNegotiationResult.success())`

marchermans replied:
Depending on the channel count, negotiation could be an expensive procedure to complete. Hence them being split.

I would leave it like this for now, but we can for sure refactor this behaviour in a new version down the line.

Technici4n commented on src/main/java/net/neoforged/neoforge/network/filters/GenericPacketSplitter.java:
```diff
@@ -207,7 +208,8 @@ public enum RemoteCompatibility {
 
     public static RemoteCompatibility getRemoteCompatibility(Connection manager) {
-        return NetworkRegistry.getInstance().isVanillaConnection(manager) ? RemoteCompatibility.ABSENT : RemoteCompatibility.PRESENT;
+        return NetworkRegistry.getInstance().isConnected(manager, ConnectionPhase.COMMON, SplitPacketPayload.ID) ?
```
I think that `COMMON` won't pickup NeoForge-negotiated channels, can you double check? `isConnected` probably needs to check that configuration AND play have the channel, right now it looks like it won't check either with COMMON.

marchermans replied:
No it does, the `ConnectionPhase.isConfiguration()` method is true for `ConnectionPhase.COMMON` and `ConnectionPhase.CONFIGURATION`. While the `ConnectionPhase.isPlay()` returns true for `ConnectionPhase.PLAY` and `ConnectionPhase.COMMON`. So this should work. Testing indicates as such as well.

Technici4n replied:
You and your confusing enums... it really doesn't make sense IMO. If a channel is supported in COMMON it should exist for both config and play. (Right now it's an `or` check.)

Since the `ConnectionPhase` is public API this should be addressed I think, at least there. `ConnectionPhase.COMMON.isPlay()` returning true seems a bit off.

marchermans replied:
I have made ConnectionPhase for now internal. We can tweak it and process it in the next few versions

Technici4n commented on src/main/java/net/neoforged/neoforge/network/connection/ConnectionPhase.java:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        if (packetListener instanceof ClientGamePacketListener) {
+            return ConnectionPhase.PLAY;
```
As a general comment, do we want to use `GAME` instead? I think "play" is yarn. We can keep play internally cause the diff might be too large otherwise, but for modder-facing APIs it would probably be better to use "game". What do you think?

marchermans replied:
It is actually called `PLAY` in ConnectionProtocol

marchermans committed:
Address some error text changes and fix the dedicated server not starting.

marchermans committed:
Rename ConnectionPhase.COMMON to ConnectionPhase.ANY.
Apply spotless.

marchermans committed:
Make ConnectionPhase internal.

Technici4n reviewed (approved):
I think we're good.

marchermans committed:
Merge branch '1.20.x' into hotfix/1.20.x/networking

XFactHD reviewed (approved):
Looks good to me :shipit:
